compiler/rustc_typeck/src/collect.rs

```diff
@@ -2647,6 +2647,91 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
                     }
                 }
             }
+        } else if tcx.sess.check_name(attr, sym::instruction_set) {
+            codegen_fn_attrs.instruction_set = match attr.meta().map(|i| i.kind) {
+                Some(MetaItemKind::List(ref items)) => match items.as_slice() {
+                    [NestedMetaItem::MetaItem(set)] => {
+                        if set
+                            .path
+                            .segments
+                            .iter()
+                            .map(|x| x.ident.name)
+                            .eq([sym::arm, sym::a32].iter().cloned())
+                        {
+                            if tcx.sess.target.target.options.has_thumb_interworking {
+                                Some(InstructionSetAttr::ArmA32)
+                            } else {
+                                struct_span_err!(
+                                    tcx.sess.diagnostic(),
+                                    attr.span,
+                                    E0779,
+                                    "target does not support `#[instruction_set]`"
+                                )
+                                .emit();
+                                None
+                            }
+                        } else if set.path.segments.iter().map(|x| x.ident.name).eq([
+                            sym::arm,
+                            sym::t32,
+                        ]
+                        .iter()
+                        .cloned())
+                        {
+                            if tcx.sess.target.target.options.has_thumb_interworking {
+                                Some(InstructionSetAttr::ArmT32)
+                            } else {
+                                struct_span_err!(
+                                    tcx.sess.diagnostic(),
+                                    attr.span,
+                                    E0779,
+                                    "target does not support `#[instruction_set]`"
+                                )
+                                .emit();
+                                None
+                            }
```

Comment:
It would be good to clean the parsing logic up a bit so that this error doesn't have to be emitted from 2 places.

Matching the argument path could also be simplified by collecting the path segments into a `Vec`.